File: android/xraymobile/src/main/java/org/xrayrust/mobile/XrayProfileImporter.kt
package org.xrayrust.mobile

import org.json.JSONArray
import org.json.JSONObject
import java.nio.ByteBuffer
import java.nio.charset.CodingErrorAction

/** Supported source formats, independent of the resulting JSON outbound name. */
enum class XrayProfileFormat(val wireValue: String, internal val capability: XrayFfiCapability) {
    Hysteria2("hysteria2", XrayFfiCapability.Hysteria2Outbound),
    Wireguard("wireguard", XrayFfiCapability.WireguardOutbound),
}

fun XrayFfiInfo.supportsProfileImport(format: XrayProfileFormat): Boolean =
    version.major == 1 && version.minor >= 5 &&
        supports(XrayFfiCapability.ProfileImport) && supports(format.capability)

/** Imports text already read by the host. No file access, network or VPN startup. */
object XrayProfileImporter {
    @JvmStatic
    @JvmOverloads
    fun profile(
        text: String,
        format: XrayProfileFormat,
        name: String? = null,
        dnsServers: List<String> = emptyList(),
    ): XrayImportedProfile = importProfile(text, format, name, dnsServers,
        XrayCore.ffiInfo(), XrayCore::importProfileJson)

    internal fun importProfile(
        text: String, format: XrayProfileFormat, name: String?, dnsServers: List<String>,
        info: XrayFfiInfo, nativeImport: (ByteArray) -> ByteArray,
    ): XrayImportedProfile {
        check(info.supportsProfileImport(format)) { "native library does not support the requested profile import" }
        require(validUtf8Size(text, 64 * 1024) &&
            (name == null || validUtf8Size(name, 128)) && dnsServers.size <= 8 &&
            dnsServers.all { validUtf8Size(it, 253) }
        ) { "invalid or oversized profile import input" }
        val request = JSONObject().put("format", format.wireValue).put("text", text)
            .put("dnsServers", JSONArray(dnsServers))
        if (name != null) request.put("name", name)
        val input = request.toString().toByteArray(Charsets.UTF_8)
        try {
            require(input.size <= 256 * 1024) { "profile import request exceeds its size limit" }
            val output = nativeImport(input)
            try {
                check(output.size <= 256 * 1024) { "invalid imported profile response size" }
                // Never attach a JSON parse exception that might quote credential data.
                val profile = runCatching {
                    val textResult = Charsets.UTF_8.newDecoder().onMalformedInput(CodingErrorAction.REPORT)
                        .onUnmappableCharacter(CodingErrorAction.REPORT).decode(ByteBuffer.wrap(output))
                    val result = JSONObject(textResult.toString())
                    check(result.get("schemaVersion") == 1)
                    XrayImportedProfile(result.get("name") as String, result.get("serverAddress") as String,
                        result.get("configJSON") as String)
                }.getOrNull()
                return checkNotNull(profile) { "native library returned an invalid imported profile" }
            } finally { output.fill(0) }
        } finally { input.fill(0) }
    }

    // Validate Unicode without creating another credential byte array. The
    // default JVM encoder replaces unpaired surrogates, changing passwords.
    private fun validUtf8Size(value: String, maxBytes: Int): Boolean {
        if (value.length > maxBytes) return false
        var bytes = 0
        var index = 0
        while (index < value.length) {
            val c = value[index++]
            bytes += when {
                c.code < 0x80 -> 1
                c.code < 0x800 -> 2
                Character.isHighSurrogate(c) -> {
                    if (index == value.length || !Character.isLowSurrogate(value[index++])) return false
                    4
                }
                Character.isLowSurrogate(c) -> return false
                else -> 3
            }
            if (bytes > maxBytes) return false
        }
        return true
    }
}

File: Sources/XrayMobileAdapter/XrayProfileImporter.swift
import Foundation
import XrayAppleShared
import XrayRust

public enum XrayProfileFormat: String, Codable, Sendable {
    case hysteria2
    case wireguard

    var capability: XrayFFICapabilities {
        self == .hysteria2 ? .hysteria2Outbound : .wireguardOutbound
    }
}

public enum XrayProfileImportError: Error, LocalizedError, Equatable {
    case unavailable
    case invalidInput
    case invalidResult

    public var errorDescription: String? {
        switch self {
        case .unavailable: return "This native library does not support the requested profile import."
        case .invalidInput: return "Invalid or oversized profile import input."
        case .invalidResult: return "The native library returned an invalid imported profile."
        }
    }
}

/// The explicit configJSON property contains credentials. Store it securely.
public struct XrayImportedProfile: Decodable, Sendable, CustomStringConvertible, CustomDebugStringConvertible {
    public let name: String
    public let serverAddress: String
    public let configJSON: String
    let schemaVersion: Int

    public var description: String { "XrayImportedProfile(<redacted>)" }
    public var debugDescription: String { description }

    public func clientProfile(
        providerBundleIdentifier: String? = nil,
        hostBundleIdentifier: String? = Bundle.main.bundleIdentifier
    ) -> XrayClientProfile {
        XrayClientProfile(
            name: name,
            providerBundleIdentifier: providerBundleIdentifier ?? XrayClientProfile.defaultProviderBundleIdentifier(
                hostBundleIdentifier: hostBundleIdentifier
            ),
            serverAddress: serverAddress,
            configJSON: configJSON
        )
    }
}

public extension XrayFFIInfo {
    func supportsProfileImport(_ format: XrayProfileFormat) -> Bool {
        version.major == 1 && version.minor >= 5 && supports(.profileImport) && supports(format.capability)
    }
}

/// Imports text already read by the host. No file access, network or VPN startup.
public enum XrayProfileImporter {
    public static func profile(
        from text: String,
        format: XrayProfileFormat,
        name: String? = nil,
        dnsServers: [String] = []
    ) throws -> XrayImportedProfile {
        try importProfile(text: text, format: format, name: name, dnsServers: dnsServers,
                          info: XrayCore.ffiInfo, nativeImport: nativeImport)
    }

    // Injectable boundary verifies availability before invoking a new native API.
    static func importProfile(
        text: String, format: XrayProfileFormat, name: String?, dnsServers: [String],
        info: XrayFFIInfo, nativeImport: (Data) throws -> Data
    ) throws -> XrayImportedProfile {
        guard info.supportsProfileImport(format) else { throw XrayProfileImportError.unavailable }
        guard text.utf8.count <= 64 * 1024, dnsServers.count <= 8,
              name.map({ $0.utf8.count <= 128 }) ?? true,
              dnsServers.allSatisfy({ $0.utf8.count <= 253 }) else {
            throw XrayProfileImportError.invalidInput
        }
        var request: [String: Any] = ["format": format.rawValue, "text": text, "dnsServers": dnsServers]
        if let name { request["name"] = name }
        var data = try JSONSerialization.data(withJSONObject: request)
        defer { data.resetBytes(in: 0..<data.count) }
        guard data.count <= 256 * 1024 else { throw XrayProfileImportError.invalidInput }
        var response = try nativeImport(data)
        defer { response.resetBytes(in: 0..<response.count) }
        guard response.count <= 256 * 1024,
              let profile = try? JSONDecoder().decode(XrayImportedProfile.self, from: response),
              profile.schemaVersion == 1 else { throw XrayProfileImportError.invalidResult }
        return profile
    }

    private static func nativeImport(_ data: Data) throws -> Data {
        try data.withUnsafeBytes { raw in
            let input = raw.bindMemory(to: UInt8.self)
            var error: OpaquePointer?
            var written = 0
            func check(_ status: XrayStatus) throws {
                defer { xray_error_free(error); error = nil }
                guard status == XRAY_STATUS_OK else {
                    let message = error.flatMap { xray_error_message($0) }.map { String(cString: $0) }
                        ?? "Profile import failed."
                    throw XrayCoreError.status(code: status, message: message)
                }
            }
            try check(xray_profile_import_json(input.baseAddress, input.count, nil, 0, &written, &error))
            guard written > 0, written <= 256 * 1024 else { throw XrayProfileImportError.invalidResult }
            var buffer = Data(count: written + 1)
            defer { buffer.resetBytes(in: 0..<buffer.count) }
            let status = buffer.withUnsafeMutableBytes { output in
                xray_profile_import_json(input.baseAddress, input.count,
                    output.bindMemory(to: CChar.self).baseAddress, output.count, &written, &error)
            }
            try check(status)
            guard written < buffer.count else { throw XrayProfileImportError.invalidResult }
            return buffer.prefix(written)
        }
    }
}
